pubsub-mapped-api/src/main/java/com/google/pubsub/clients/producer/KafkaProducer.java

```diff
@@ -0,0 +1,398 @@
+              .setInitialRpcTimeout(Duration.ofMillis(timeout))
+              .setMaxRpcTimeout(Duration.ofMillis((retries + 1) * timeout))
+
+              .setTotalTimeout(Duration.ofMillis((retries + 2) * timeout))
+              .build())
+
+          .build();
+    } catch (IOException e) {
+      throw new KafkaException("Failed to construct kafka producer", e);
+    }
+    return pub;
+  }
+
+  /**
+   * Sends the given record.
+   */
+  public Future<RecordMetadata> send(ProducerRecord<K, V> originalRecord) {
+    return send(originalRecord, null);
+  }
+
+  //TODO: mimic all kafka's exceptions.
+  //TODO: there is an onSendError() in the interceptors, while there are no errors.
+
+  /**
+   * Sends the given record and invokes the specified callback.
+   * The given record must have the same topic as the producer.
+   */
+  public Future<RecordMetadata> send(ProducerRecord<K, V> originalRecord, Callback callback) {
+    if (closed.get())
+      throw new IllegalStateException("Cannot send after the producer is closed.");
+
+    ProducerRecord<K, V> record =
+        this.interceptors == null ? originalRecord : this.interceptors.onSend(originalRecord);
+
+    if (record == null)
+      throw new NullPointerException();
+
+    final Callback cb = callback;
+
+    byte[] valueBytes = null;
+    if (record.value() != null) {
+      try {
+        valueBytes = valueSerializer.serialize(record.topic(), record.value());
+      } catch (ClassCastException e) {
+        throw new SerializationException("Can't convert value of class " +
+            record.value().getClass().getName() + " to class " +
+            producerConfig.getClass(ProducerConfig.VALUE_SERIALIZER_CLASS_CONFIG).getName() +
+            " specified in value.serializer");
+      }
+    }
+
+    if (valueBytes == null || valueBytes.length == 0)
+      throw new NullPointerException("Value cannot be null or an empty string.");
+
+    Map<String, String> attributes = new HashMap<>();
+
+    attributes.put("id", clientId);
+
+    byte[] keyBytes = null;
+    if (record.key() != null) {
+      try {
+        keyBytes = keySerializer.serialize(record.topic(), record.key());
+      } catch (ClassCastException cce) {
+        throw new SerializationException("Can't convert key of class " +
+            record.key().getClass().getName() + " to class " +
+            producerConfig.getClass(ProducerConfig.KEY_SERIALIZER_CLASS_CONFIG).getName() +
+            " specified in key.serializer");
+      }
+    }
+    attributes.put("key", keyBytes == null ? "" : new String(Base64.encodeBase64(keyBytes)));
+
+    PubsubMessage msg = PubsubMessage.newBuilder()
+        .setData(ByteString.copyFrom(valueBytes))
+        .putAllAttributes(attributes)
+        .build();
+
+    ApiFuture<String> messageIdFuture = null;
+
+    messageIdFuture =
+        publishers.computeIfAbsent(
+            record.topic(), K -> new AtomicReference<>(createPublisher(record.topic())))
```

Comment:
You can rename "K" to "topic" and then just use it in place of record.topic(). Is the AtomicReference necessary?